modules/basegl/src/processors/volumeprocessing/volumeregionshrink.cpp

```diff
@@ -0,0 +1,172 @@
+    Tags::GL,                         // Tags
+};
+const ProcessorInfo VolumeRegionShrink::getProcessorInfo() const { return processorInfo_; }
+
+namespace {
+
+constexpr std::string_view fragStr = R"(
+#include "utils/structs.glsl"
+#include "utils/sampler3d.glsl"
+
+uniform sampler3D volume;
+uniform VolumeParameters volumeParameters;
+
+in vec4 texCoord_;
+
+void main() {
+    vec4 value = getVoxel(volume, volumeParameters, texCoord_.xyz);
+
+    bool edge = false;
+    for(int z = -1; z < 1; z++) {
+        for(int y = -1; y < 1; y++) {
+            for(int x = -1; x < 1; x++){
+
+                vec4 neighbor = getVoxel(volume, volumeParameters, texCoord_.xyz + vec3(x,y,z)*volumeParameters.reciprocalDimensions);
+                if (neighbor != value) {
+                    edge = true;
+                } 
+            }
+        }
+    }
+    if (edge) {
+        FragData0 = vec4(0.0, 0.0, 0.0, 0.0);
+    } else {
+        FragData0 = value;
+    }
+}   
+)";
+
+}
+
+VolumeRegionShrink::VolumeRegionShrink()
+    : Processor()
+    , inport_{"inputVolume"}
+    , outport_{"outputVolume"}
+    , iterations_{"iterations", "Iterations", 3, 0, 25}
+    , fragShader_{std::make_shared<StringShaderResource>("VolumeRegionShrink.frag", fragStr)}
+    , shader_({{ShaderType::Vertex, utilgl::findShaderResource("volume_gpu.vert")},
+               {ShaderType::Geometry, utilgl::findShaderResource("volume_gpu.geom")},
+               {ShaderType::Fragment, fragShader_}},
+              Shader::Build::Yes) {
+
+    addPort(inport_);
+    addPort(outport_);
+
+    addProperty(iterations_);
+
+    shader_.onReload([this]() { invalidate(InvalidationLevel::InvalidResources); });
+}
+
+void VolumeRegionShrink::process() {
+    auto volume = inport_.getData();
+
+    if (iterations_ == 0) {
+        outport_.setData(volume);
+        return;
+    }
+
+    if (!out_[0] || out_[0]->getDataFormat() != volume->getDataFormat() ||
+        out_[0]->getDimensions() != volume->getDimensions()) {
+        out_[0] = std::shared_ptr<Volume>(volume->clone());
+    }
+
+    out_[0]->setModelMatrix(volume->getModelMatrix());
+    out_[0]->setWorldMatrix(volume->getWorldMatrix());
+    out_[0]->copyMetaDataFrom(*volume);
+    out_[0]->dataMap_ = volume->dataMap_;
+
+    const size3_t dim{volume->getDimensions()};
+    glViewport(0, 0, static_cast<GLsizei>(dim.x), static_cast<GLsizei>(dim.y));
+
+    fbo_.activate();
+    VolumeGL* outGL0 = out_[0]->getEditableRepresentation<VolumeGL>();
+    fbo_.attachColorTexture(outGL0->getTexture().get(), 0);
+    out_[0]->invalidateHistogram();
+
+    utilgl::Activate as{&shader_};
+
+    // Iteration 1
+    {
+        glDrawBuffer(GL_COLOR_ATTACHMENT0);
+        TextureUnitContainer cont;
+        utilgl::bindAndSetUniforms(shader_, cont, *volume, "volume");
+        utilgl::multiDrawImagePlaneRect(static_cast<int>(dim.z));
+    }
+    if (iterations_ == 1) {
+        outport_.setData(out_[0]);
```

Comment:
Deactivate FBO?

Reply:
Also remove the second temporary volume in case it exist?

Reply:
I want to avoid having to recreate it if it needs to recompute. guess that is a space / time trade off though.

Reply:
Recreation would only happen if the property changes so it should not be too often. But yeah, a trad off

Reply:
of if the input changes, which can happen often in some cases...

Reply:
It would only deallocate once if iterations_ == 1. Then reallocate when iterations_ changes

Reply:
ahh now I understand what you mean, i missed the fact that your comment was in the iterations ==1 block, yes that would make sense.